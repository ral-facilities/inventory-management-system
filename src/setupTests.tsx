// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import '@testing-library/jest-dom';
import { RenderOptions, render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import {
  CatalogueCategory,
  CatalogueCategoryFormData,
  CatalogueItem,
  Item,
  Manufacturer,
} from './app.types';
import CatalogueCategoryJSON from './mocks/CatalogueCategory.json';
import CatalogueItemJSON from './mocks/CatalogueItems.json';
import ManufacturerJSON from './mocks/manufacturer.json';
import ItemsJSON from './mocks/Items.json';
import { server } from './mocks/server';
import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import enGB from 'date-fns/locale/en-GB';

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
    logger: {
      log: console.log,
      warn: console.warn,
      error: jest.fn(),
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

export function renderComponentWithBrowserRouter(
  ui: React.ReactElement,
  {
    // Automatically create a store instance if no store was passed i
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <BrowserRouter>
        <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
          <QueryClientProvider client={queryClient}>
            {children}
          </QueryClientProvider>
        </LocalizationProvider>
      </BrowserRouter>
    );
  }
  return {
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export function renderComponentWithMemoryRouter(
  ui: React.ReactElement,
  path: string,
  {
    // Automatically create a store instance if no store was passed i
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <MemoryRouter initialEntries={[path]}>
        <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
          <QueryClientProvider client={queryClient}>
            {children}
          </QueryClientProvider>
        </LocalizationProvider>
      </MemoryRouter>
    );
  }
  return {
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export const hooksWrapperWithProviders = (queryClient?: QueryClient) => {
  const testQueryClient = queryClient ?? createTestQueryClient();
  const wrapper = ({ children }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  );
  return wrapper;
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryFormData[] => {
  const filteredCategories = CatalogueCategoryJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].catalogue_item_properties ?? [];
  return properties;
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoryJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemJSON.filter(
    (catalogueitem) => catalogueitem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturerJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};
