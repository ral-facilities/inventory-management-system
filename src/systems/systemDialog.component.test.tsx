import { fireEvent, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import axios from 'axios';
import React from 'react';
import { renderComponentWithBrowserRouter } from '../setupTests';
import SystemDialog, { SystemDialogProps } from './systemDialog.component';
import { System, SystemImportanceType } from '../app.types';

describe('Systems Dialog', () => {
  let props: SystemDialogProps;
  let user;
  let axiosPostSpy;
  let axiosPatchSpy;

  const mockOnClose = jest.fn();

  const createView = () => {
    return renderComponentWithBrowserRouter(<SystemDialog {...props} />);
  };

  // Modifies values when given a value that is not undefined
  const modifyValues = (values: {
    name?: string;
    description?: string;
    location?: string;
    owner?: string;
    importance?: SystemImportanceType;
  }) => {
    values.name !== undefined &&
      fireEvent.change(screen.getByLabelText('Name *'), {
        target: { value: values.name },
      });
    values.description !== undefined &&
      fireEvent.change(screen.getByLabelText('Description'), {
        target: { value: values.description },
      });
    values.location !== undefined &&
      fireEvent.change(screen.getByLabelText('Location'), {
        target: { value: values.location },
      });
    values.owner !== undefined &&
      fireEvent.change(screen.getByLabelText('Owner'), {
        target: { value: values.owner },
      });

    if (values.importance !== undefined) {
      fireEvent.mouseDown(screen.getByLabelText('Importance'));
      fireEvent.click(
        within(screen.getByRole('listbox')).getByText(values.importance)
      );
    }
  };

  beforeEach(() => {
    props = {
      open: true,
      onClose: mockOnClose,
      parentId: null,
      type: 'add',
    };
    user = userEvent.setup();
    axiosPostSpy = jest.spyOn(axios, 'post');
    axiosPatchSpy = jest.spyOn(axios, 'patch');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Add', () => {
    beforeEach(() => {
      props.type = 'add';
      props.parentId = null;
    });

    it('renders correctly when adding', async () => {
      const view = createView();

      expect(screen.getByText('Add System')).toBeInTheDocument();
      expect(view.baseElement).toMatchSnapshot();
    });

    it('renders correctly when adding a subsystem', async () => {
      props.parentId = 'parent-id';

      createView();

      expect(screen.getByText('Add Subsystem')).toBeInTheDocument();
    });

    it('calls onClose when cancel is clicked', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Cancel' }));

      expect(mockOnClose).toHaveBeenCalled();
    });

    it('adds a system', async () => {
      props.parentId = 'parent-id';

      createView();

      const values = {
        name: 'System name',
        description: 'System description',
        location: 'System location',
        owner: 'System owner',
        importance: SystemImportanceType.HIGH,
      };
      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(axiosPostSpy).toHaveBeenCalledWith('/v1/systems', {
        ...values,
        parent_id: 'parent-id',
      });

      expect(mockOnClose).toHaveBeenCalled();
    });

    it('adds a system with only mandatory fields', async () => {
      createView();

      const values = {
        name: 'System name',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(axiosPostSpy).toHaveBeenCalledWith('/v1/systems', {
        ...values,
        importance: SystemImportanceType.MEDIUM,
      });
      expect(mockOnClose).toHaveBeenCalled();
    });

    it('displays error message when name field is not filled in', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(screen.getByText('Please enter a name')).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('displays error message when attempting to add a system with a duplicate name', async () => {
      createView();

      const values = {
        name: 'Error 409',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(
        screen.getByText(
          'A System with the same name already exists within the same parent System'
        )
      ).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('displays error message when an unknown error occurs', async () => {
      createView();

      const values = {
        name: 'Error 500',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(
        screen.getByText('Please refresh and try again')
      ).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });
  });

  describe('Edit', () => {
    const MOCK_SELECTED_SYSTEM: System = {
      name: 'Mock laser',
      location: 'Location',
      owner: 'Owner',
      importance: SystemImportanceType.HIGH,
      description: 'Description',
      parent_id: null,
      id: '65328f34a40ff5301575a4e3',
      code: 'mock-laser',
    };

    beforeEach(() => {
      props.type = 'edit';
      props.parentId = undefined;
      props.selectedSystem = MOCK_SELECTED_SYSTEM;
    });

    it('renders correctly when editing', async () => {
      createView();

      expect(screen.getByText('Edit System')).toBeInTheDocument();
    });

    it('calls onClose when cancel is clicked', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Cancel' }));

      expect(mockOnClose).toHaveBeenCalled();
    });

    it('edits a system', async () => {
      createView();

      const values = {
        name: 'System name',
        description: 'System description',
        location: 'System location',
        owner: 'System owner',
        importance: SystemImportanceType.LOW,
      };
      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(axiosPatchSpy).toHaveBeenCalledWith(
        `/v1/systems/${MOCK_SELECTED_SYSTEM['id']}`,
        values
      );

      expect(mockOnClose).toHaveBeenCalled();
    });

    it('edits a system removing non-manditory fields', async () => {
      createView();

      const values = {
        description: '',
        location: '',
        owner: '',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(axiosPatchSpy).toHaveBeenCalledWith(
        `/v1/systems/${MOCK_SELECTED_SYSTEM['id']}`,
        { description: null, location: null, owner: null }
      );
      expect(mockOnClose).toHaveBeenCalled();
    });

    it('edits only a systems name', async () => {
      createView();

      const values = {
        name: 'System name',
      };
      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(axiosPatchSpy).toHaveBeenCalledWith(
        `/v1/systems/${MOCK_SELECTED_SYSTEM['id']}`,
        values
      );

      expect(mockOnClose).toHaveBeenCalled();
    });

    it('displays error message when no fields have been changed that disappears when the form is modified', async () => {
      createView();

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(
        screen.getByText('Please edit a form entry before clicking save')
      ).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();

      modifyValues({ description: 'New description' });

      expect(
        screen.queryByText('Please edit a form entry before clicking save')
      ).not.toBeInTheDocument();
    });

    it('displays error message when name field is not filled in', async () => {
      createView();

      modifyValues({ name: '' });

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(screen.getByText('Please enter a name')).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('displays error message when attempting to edit a system with a duplicate name', async () => {
      createView();

      const values = {
        name: 'Error 409',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(
        screen.getByText(
          'A System with the same name already exists within the same parent System'
        )
      ).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });

    it('displays error message when an unknown error occurs', async () => {
      createView();

      const values = {
        name: 'Error 500',
      };

      modifyValues(values);

      await user.click(screen.getByRole('button', { name: 'Save' }));

      expect(
        screen.getByText('Please refresh and try again')
      ).toBeInTheDocument();
      expect(mockOnClose).not.toHaveBeenCalled();
    });
  });
});
