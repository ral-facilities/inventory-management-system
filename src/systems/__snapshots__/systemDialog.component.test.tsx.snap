// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Systems Dialog > Add > renders correctly when adding 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-1424xw8-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-1cennmq-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-19do60a-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        aria-modal="true"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth css-1peo4km-MuiPaper-root-MuiDialog-paper"
        role="dialog"
        style="--Paper-shadow: 0px 11px 15px -7px rgba(0,0,0,0.2),0px 24px 38px 3px rgba(0,0,0,0.14),0px 9px 46px 8px rgba(0,0,0,0.12);"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-rvoa5x-MuiTypography-root-MuiDialogTitle-root"
          id=":r0:"
        >
          Add System
        </h2>
        <div
          class="MuiDialogContent-root css-kw13he-MuiDialogContent-root"
        >
          <div
            class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-column MuiGrid2-spacing-xs-2 css-1eaccnh-MuiGrid2-root"
          >
            <div
              class="MuiGrid2-root MuiGrid2-direction-xs-row css-53ts4k-MuiGrid2-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="system-name-input"
                  id="system-name-input-label"
                >
                  Name
                  <span
                    aria-hidden="true"
                    class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                  >
                     
                    *
                  </span>
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                    id="system-name-input"
                    name="name"
                    required=""
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-81qg8w"
                    >
                      <span>
                        Name
                         
                        *
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid2-root MuiGrid2-direction-xs-row css-1fzlhpv-MuiGrid2-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="system-description-input"
                  id="system-description-input-label"
                >
                  Description
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-zsp0bh-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
                    id="system-description-input"
                    name="description"
                    style="height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-w4nesw-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-81qg8w"
                    >
                      <span>
                        Description
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid2-root MuiGrid2-direction-xs-row css-1fzlhpv-MuiGrid2-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="system-location-input"
                  id="system-location-input-label"
                >
                  Location
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                    id="system-location-input"
                    name="location"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-81qg8w"
                    >
                      <span>
                        Location
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid2-root MuiGrid2-direction-xs-row css-1fzlhpv-MuiGrid2-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-19qnlrw-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for="system-owner-input"
                  id="system-owner-input-label"
                >
                  Owner
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1blp12k-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-16wblaj-MuiInputBase-input-MuiOutlinedInput-input"
                    id="system-owner-input"
                    name="owner"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-81qg8w"
                    >
                      <span>
                        Owner
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid2-root MuiGrid2-direction-xs-row css-1fzlhpv-MuiGrid2-root"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-o5oj2-MuiAutocomplete-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="true"
                    for="importance-select"
                    id="importance-select-label"
                  >
                    Importance
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <div
                      class="MuiChip-root MuiChip-filled MuiChip-sizeMedium MuiChip-colorDefault MuiChip-filledDefault css-g92mod-MuiChip-root"
                    >
                      <span
                        class="MuiChip-label MuiChip-labelMedium css-1dybbl5-MuiChip-label"
                      >
                        medium
                      </span>
                    </div>
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
                      id="importance-select"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-w1u3ce"
                      >
                        <span>
                          Importance
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-1beia2c-MuiDialogActions-root"
        >
          <div
            class="MuiBox-root css-ldo4d5"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-s8qwki-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Cancel
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-s8qwki-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Save
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
